File: Task/Sealed-classes-and-methods/Go/sealed-classes-and-methods.go
package main

import "fmt"

type typeid int

const (
    PARENT typeid = iota
    CHILD
)

type parent struct {
    name string
    age  int
}

type child struct {
    parent // embedded struct
}

func (p *parent) watchMovie(id typeid) {
    if id == CHILD && p.age < 15 {
        fmt.Printf("Sorry, %s, you are too young to watch the movie.\n", p.name)
    } else {
        fmt.Printf("%s is watching the movie...\n", p.name)
    }
}

func main() {
    p := &parent{"Donald", 42}
    p.watchMovie(PARENT)
    c1 := &child{parent{"Lisa", 18}}
    c2 := &child{parent{"Fred", 10}}
    c1.watchMovie(CHILD)
    c2.watchMovie(CHILD)
}

File: Task/Sealed-classes-and-methods/C/sealed-classes-and-methods.c
#include <stdio.h>

typedef enum {
    PARENT,
    CHILD
} typeid;

typedef struct {
    const char* name;
    int age;
} parent;

typedef struct {
    parent p;
} child;

void watchMovie(parent *p, typeid id) {
    if (id == CHILD && p->age < 15) {
        printf("Sorry, %s, you are too young to watch the movie.\n", p->name);
    } else {
        printf("%s is watching the movie...\n", p->name);
    }
}
int main() {
    parent p = { "Donald", 42 };
    child c1 = { "Lisa", 18 };
    child c2 = { "Fred", 10 };
    watchMovie(&p, PARENT);
    watchMovie(&c1.p, CHILD);
    watchMovie(&c2.p, CHILD);
    return 0;
}
